File: packages/core/src/air/parser.py
from __future__ import annotations

from pathlib import Path
import xml.etree.ElementTree as ET

from .model import (
    AnalogSubsystem,
    Bridge,
    Component,
    ExportTarget,
    FirmwareBinding,
    FirmwareProject,
    FirmwareTask,
    Interface,
    Metadata,
    Net,
    PinConnection,
    PowerDomain,
    Probe,
    SimulationProfile,
    SystemIR,
    Test,
)
from .normalizer import normalize_air_tree


def parse_file(path: Path) -> tuple[SystemIR, ET.ElementTree]:
    tree = ET.parse(path)
    return parse_tree(tree), tree


def parse_string(xml_text: str) -> tuple[SystemIR, ET.ElementTree]:
    root = ET.fromstring(xml_text)
    tree = ET.ElementTree(root)
    return parse_tree(tree), tree


def parse_tree(tree: ET.ElementTree) -> SystemIR:
    tree = normalize_air_tree(tree)
    root = tree.getroot()
    if root.tag != "system":
        raise ValueError("AIR document root must be <system>")

    metadata_el = root.find("metadata")
    metadata = Metadata(
        title=_text(metadata_el, "title"),
        description=_text(metadata_el, "description"),
        author=_text(metadata_el, "author"),
        created_at=_text(metadata_el, "created_at"),
    )

    nets = {
        net.attrib["id"]: Net(
            id=net.attrib["id"],
            role=net.attrib.get("role", ""),
            nominal_voltage=net.attrib.get("nominal_voltage"),
        )
        for net in root.findall("./nets/net")
        if "id" in net.attrib
    }

    power_domains = {
        domain.attrib["id"]: PowerDomain(
            id=domain.attrib["id"],
            net=domain.attrib.get("net", ""),
            nominal=domain.attrib.get("nominal"),
            source=domain.attrib.get("source"),
        )
        for domain in root.findall("./power_domains/domain")
        if "id" in domain.attrib
    }

    components: dict[str, Component] = {}
    for element in root.findall("./components/component"):
        component_id = element.attrib.get("id", "")
        pins = {
            pin.attrib.get("name", ""): PinConnection(
                name=pin.attrib.get("name", ""),
                net=pin.attrib.get("net") or pin.attrib.get("node") or pin.attrib.get("ref") or "",
                function=pin.attrib.get("function"),
            )
            for pin in element.findall("pin")
        }
        properties = {
            prop.attrib.get("name", ""): prop.attrib.get("value", "")
            for prop in element.findall("property")
        }
        value_el = element.find("value")
        components[component_id] = Component(
            id=component_id,
            type=element.attrib.get("type", ""),
            part=element.attrib.get("part"),
            spice_model=element.attrib.get("spice_model"),
            spice_subckt=element.attrib.get("spice_subckt"),
            value=value_el.text.strip() if value_el is not None and value_el.text else None,
            pins=pins,
            properties=properties,
        )

    analog = []
    for subsystem in root.findall("./analog/subsystem"):
        analog.append(
            AnalogSubsystem(
                id=subsystem.attrib.get("id", ""),
                uses=[u.attrib.get("component", "") for u in subsystem.findall("uses")],
                probes=[
                    Probe(
                        id=p.attrib.get("id", ""),
                        net=p.attrib.get("net", ""),
                        quantity=p.attrib.get("quantity", ""),
                    )
                    for p in subsystem.findall("probe")
                ],
            )
        )

    interfaces = {}
    for iface in root.findall("./interfaces/interface"):
        data: dict[str, object] = {}
        for child in iface:
            value = dict(child.attrib)
            existing = data.get(child.tag)
            if existing is None:
                data[child.tag] = value
            elif isinstance(existing, list):
                existing.append(value)
            else:
                data[child.tag] = [existing, value]
        interfaces[iface.attrib.get("id", "")] = Interface(
            id=iface.attrib.get("id", ""),
            type=iface.attrib.get("type", ""),
            data=data,
        )

    firmware_projects = {}
    firmware_bindings = {}
    firmware_tasks = {}
    firmware_el = root.find("firmware")
    if firmware_el is not None:
        for project in firmware_el.findall("project"):
            project_id = project.attrib.get("id", "")
            firmware_projects[project_id] = FirmwareProject(
                id=project_id,
                target=project.get("target", ""),
                framework=project.get("framework", ""),
                language=project.get("language", ""),
                board=_text(project, "board"),
                source_tree=project.get("source_tree", ""),
            )
        for binding in firmware_el.findall("binding"):
            binding_id = binding.attrib.get("id", "")
            firmware_bindings[binding_id] = FirmwareBinding(
                id=binding_id,
                signal=_child_attr(binding, "signal", "name"),
                component=_child_attr(binding, "component", "ref"),
                peripheral=_text(binding, "peripheral"),
                channel=_text(binding, "channel"),
                net=_text(binding, "net"),
            )
        for task in firmware_el.findall("task"):
            task_id = task.attrib.get("id", "")
            operations = [
                {**{"op": child.tag}, **dict(child.attrib), **({"text": child.text.strip()} if child.text and child.text.strip() else {})}
                for child in task
                if child.tag != "period"
            ]
            firmware_tasks[task_id] = FirmwareTask(
                id=task_id,
                target=task.attrib.get("target", ""),
                period=_text(task, "period"),
                operations=operations,
            )

    bridges = []
    for b in root.findall("./bridges/bridge"):
        bridge_data = dict(b.attrib)
        for child in b:
            bridge_data[child.tag] = child.attrib
        bridges.append(Bridge(id=b.attrib.get("id", ""), type=b.attrib.get("type", ""), data=bridge_data))

    tests = {}
    for test in root.findall("./tests/test"):
        test_id = test.attrib.get("id", "")
        setup = {}
        for child in test.findall("./setup/*"):
            if child.tag == "set_voltage":
                setup[child.attrib.get("net", child.tag)] = child.attrib.get("value", "")
            elif child.tag == "set_current":
                setup[f"current:{child.attrib.get('component', '')}"] = child.attrib.get("value", "")
            elif child.tag == "load_step":
                component = child.attrib.get("component", "")
                setup[f"load_step:{component}"] = ",".join(
                    [
                        child.attrib.get("from", ""),
                        child.attrib.get("to", ""),
                        child.attrib.get("at", "0s"),
                        child.attrib.get("rise", "1us"),
                    ]
                )
            else:
                setup[child.attrib.get("net", child.tag)] = child.attrib.get("value", "")
        assertions = [dict(assertion.attrib, op=assertion.tag) for assertion in test if assertion.tag.startswith("assert_")]
        run = test.find("run")
        tests[test_id] = Test(
            id=test_id,
            description=_text(test, "description"),
            setup=setup,
            duration=run.attrib.get("duration", "") if run is not None else "",
            assertions=assertions,
        )

    profiles = {}
    for profile in root.findall("./simulation_profiles/profile"):
        profile_id = profile.attrib.get("id", "")
        props = {}
        for p in profile.findall("property"):
            name = p.attrib.get("name")
            if name:
                props[name] = p.attrib.get("value", "")
        profiles[profile_id] = SimulationProfile(
            id=profile_id,
            default=profile.attrib.get("default", "false").lower() == "true",
            backends=[b.attrib.get("type", "") for b in profile.findall("backend")],
            included_subsystems=[i.attrib.get("subsystem", "") for i in profile.findall("include")],
            tests=[r.attrib.get("test", "") for r in profile.findall("run")],
            properties=props,
        )

    exports = [
        ExportTarget(
            target=e.attrib.get("target", ""),
            enabled=e.attrib.get("enabled", "false").lower() == "true",
        )
        for e in root.findall("./exports/export")
    ]

    return SystemIR(
        name=root.attrib.get("name", ""),
        ir_version=root.attrib.get("ir_version", ""),
        metadata=metadata,
        nets=nets,
        power_domains=power_domains,
        components=components,
        interfaces=interfaces,
        analog=analog,
        firmware_projects=firmware_projects,
        firmware_bindings=firmware_bindings,
        firmware_tasks=firmware_tasks,
        bridges=bridges,
        tests=tests,
        simulation_profiles=profiles,
        exports=exports,
    )


def _text(parent: ET.Element | None, tag: str) -> str:
    if parent is None:
        return ""
    child = parent.find(tag)
    return child.text.strip() if child is not None and child.text else ""


def _child_attr(parent: ET.Element, tag: str, attr: str) -> str:
    child = parent.find(tag)
    return child.attrib.get(attr, "") if child is not None else ""

File: packages/air-ts/src/parser.ts
/**
 * Port of `packages/core/src/air/parser.py` (parse_tree).
 *
 * Consumes a normalized element tree and builds the typed SystemIR. Every
 * leniency the oracle implements is reproduced:
 *   - pin `net` falls back to `node`, then `ref` (net= / node= / ref= aliases);
 *   - net-owned `<node>` pins are materialized by the normalizer (see
 *     normalizer.ts) into component-owned `<pin>`s before we read them here;
 *   - implied roles / value-from-parameters likewise come from the normalizer.
 *
 * PARITY: the attribute names, the `.strip()` on <value>, the empty-string
 * fallbacks, the firmware `source_tree`/`board` attribute-vs-child asymmetry,
 * the test-setup special cases (set_voltage / set_current / load_step), and the
 * interface/bridge `data` shapes are copied verbatim from parser.py.
 */

import {
  type XmlElement,
  childElements,
  findAll,
  find,
  attr,
  hasAttr,
  elementText,
} from "./xml.js";
import { normalizeTree } from "./normalizer.js";
import type {
  AnalogSubsystem,
  Bridge,
  BridgeDatum,
  Component,
  ExportTarget,
  FirmwareBinding,
  FirmwareOperation,
  FirmwareProject,
  FirmwareTask,
  Interface,
  InterfaceDatum,
  Metadata,
  Net,
  PinConnection,
  PowerDomain,
  Probe,
  SimulationProfile,
  SystemIR,
  Test,
} from "./model.js";

/** Raised when the AIR document root is not <system> (mirrors ValueError). */
export class AirParseError extends Error {
  constructor(message: string) {
    super(message);
    this.name = "AirParseError";
  }
}

/**
 * Build a SystemIR from a raw (un-normalized) element tree. Normalization is
 * applied internally on a clone; the input `root` is left untouched so the
 * caller can canonicalize the raw tree (matching parse_file's semantics, where
 * `parse_tree` rebinds `tree` locally and the caller keeps the raw tree).
 */
export function parseTree(rawRoot: XmlElement): SystemIR {
  const root = normalizeTree(rawRoot);
  if (root.tag !== "system") {
    throw new AirParseError("AIR document root must be <system>");
  }

  const metadataEl = find(root, "metadata");
  const metadata: Metadata = {
    title: childText(metadataEl, "title"),
    description: childText(metadataEl, "description"),
    author: childText(metadataEl, "author"),
    created_at: childText(metadataEl, "created_at"),
  };

  // All keyed collections are Maps: insertion (document) order for every key
  // type, mirroring Python dicts. Plain objects would iterate integer-like keys
  // (pin names "1"/"2", numeric ids) numerically first -- observable in
  // validation's emission order and pins[0] access (issue #8 rework round 1;
  // see model.ts header). Map.set matches Python dict assignment on duplicate
  // keys: last value wins, first insertion position kept.
  const nets = new Map<string, Net>();
  for (const net of findAllPath(root, "nets", "net")) {
    if (!hasAttr(net, "id")) continue;
    const id = attr(net, "id");
    nets.set(id, {
      id,
      role: attr(net, "role", ""),
      nominal_voltage: net.attrib.has("nominal_voltage")
        ? attr(net, "nominal_voltage")
        : null,
    });
  }

  const powerDomains = new Map<string, PowerDomain>();
  for (const domain of findAllPath(root, "power_domains", "domain")) {
    if (!hasAttr(domain, "id")) continue;
    const id = attr(domain, "id");
    powerDomains.set(id, {
      id,
      net: attr(domain, "net", ""),
      nominal: domain.attrib.has("nominal") ? attr(domain, "nominal") : null,
      source: domain.attrib.has("source") ? attr(domain, "source") : null,
    });
  }

  const components = new Map<string, Component>();
  for (const element of findAllPath(root, "components", "component")) {
    const componentId = attr(element, "id", "");
    const pins = new Map<string, PinConnection>();
    for (const pin of findAll(element, "pin")) {
      const name = attr(pin, "name", "");
      // net = net || node || ref || "" (Python `or`-chain on attribute values).
      const net =
        orChain(
          pin.attrib.get("net"),
          pin.attrib.get("node"),
          pin.attrib.get("ref"),
        ) ?? "";
      pins.set(name, {
        name,
        net,
        function: pin.attrib.has("function") ? attr(pin, "function") : null,
      });
    }
    const properties = new Map<string, string>();
    for (const prop of findAll(element, "property")) {
      properties.set(attr(prop, "name", ""), attr(prop, "value", ""));
    }
    const valueEl = find(element, "value");
    const valueText =
      valueEl !== null ? elementText(valueEl) : "";
    components.set(componentId, {
      id: componentId,
      type: attr(element, "type", ""),
      part: element.attrib.has("part") ? attr(element, "part") : null,
      spice_model: element.attrib.has("spice_model")
        ? attr(element, "spice_model")
        : null,
      spice_subckt: element.attrib.has("spice_subckt")
        ? attr(element, "spice_subckt")
        : null,
      // `value_el.text.strip()` only when there IS non-empty text; else None.
      value: valueEl !== null && valueText ? valueText.trim() : null,
      pins,
      properties,
    });
  }

  const analog: AnalogSubsystem[] = [];
  for (const subsystem of findAllPath(root, "analog", "subsystem")) {
    const probes: Probe[] = findAll(subsystem, "probe").map((p) => ({
      id: attr(p, "id", ""),
      net: attr(p, "net", ""),
      quantity: attr(p, "quantity", ""),
    }));
    analog.push({
      id: attr(subsystem, "id", ""),
      uses: findAll(subsystem, "uses").map((u) => attr(u, "component", "")),
      probes,
    });
  }

  const interfaces = new Map<string, Interface>();
  for (const iface of findAllPath(root, "interfaces", "interface")) {
    const data: Record<string, InterfaceDatum> = {};
    for (const child of childElements(iface)) {
      const value = attribMap(child);
      const existing = data[child.tag];
      if (existing === undefined) {
        data[child.tag] = value;
      } else if (Array.isArray(existing)) {
        existing.push(value);
      } else {
        data[child.tag] = [existing, value];
      }
    }
    const id = attr(iface, "id", "");
    interfaces.set(id, { id, type: attr(iface, "type", ""), data });
  }

  const firmwareProjects = new Map<string, FirmwareProject>();
  const firmwareBindings = new Map<string, FirmwareBinding>();
  const firmwareTasks = new Map<string, FirmwareTask>();
  const firmwareEl = find(root, "firmware");
  if (firmwareEl !== null) {
    for (const project of findAll(firmwareEl, "project")) {
      const projectId = attr(project, "id", "");
      firmwareProjects.set(projectId, {
        id: projectId,
        target: attr(project, "target", ""),
        framework: attr(project, "framework", ""),
        language: attr(project, "language", ""),
        board: childText(project, "board"),
        // NOTE: read as an ATTRIBUTE. When source_tree is a child element
        // (<source_tree path="..."/>) there is no attribute, so this is "".
        source_tree: attr(project, "source_tree", ""),
      });
    }
    for (const binding of findAll(firmwareEl, "binding")) {
      const bindingId = attr(binding, "id", "");
      firmwareBindings.set(bindingId, {
        id: bindingId,
        signal: childAttr(binding, "signal", "name"),
        component: childAttr(binding, "component", "ref"),
        peripheral: childText(binding, "peripheral"),
        channel: childText(binding, "channel"),
        net: childText(binding, "net"),
      });
    }
    for (const task of findAll(firmwareEl, "task")) {
      const taskId = attr(task, "id", "");
      const operations: FirmwareOperation[] = [];
      for (const child of childElements(task)) {
        if (child.tag === "period") continue;
        const op: FirmwareOperation = { op: child.tag };
        for (const [k, v] of child.attrib) op[k] = v;
        const text = elementText(child).trim();
        if (text) op["text"] = text;
        operations.push(op);
      }
      firmwareTasks.set(taskId, {
        id: taskId,
        target: attr(task, "target", ""),
        period: childText(task, "period"),
        operations,
      });
    }
  }

  const bridges: Bridge[] = [];
  for (const b of findAllPath(root, "bridges", "bridge")) {
    const bridgeData: Record<string, BridgeDatum> = {};
    // Start with the bridge's own attributes (flat strings)...
    for (const [k, v] of b.attrib) bridgeData[k] = v;
    // ...then overlay each child tag with its attribute map.
    for (const child of childElements(b)) {
      bridgeData[child.tag] = attribMap(child);
    }
    bridges.push({
      id: attr(b, "id", ""),
      type: attr(b, "type", ""),
      data: bridgeData,
    });
  }

  const tests = new Map<string, Test>();
  for (const test of findAllPath(root, "tests", "test")) {
    const testId = attr(test, "id", "");
    const setup = new Map<string, string>();
    // PARITY (rework round 1 audit): the oracle iterates test.findall("./setup/*")
    // -- the children of EVERY <setup> section under the test, not just the
    // first. Same multi-section family as the DUPLICATE_ID scan (root cause B).
    const setupChildren: XmlElement[] = [];
    for (const setupEl of findAll(test, "setup")) {
      for (const child of childElements(setupEl)) setupChildren.push(child);
    }
    for (const child of setupChildren) {
      if (child.tag === "set_voltage") {
        setup.set(attr(child, "net", child.tag), attr(child, "value", ""));
      } else if (child.tag === "set_current") {
        setup.set(`current:${attr(child, "component", "")}`, attr(
          child,
          "value",
          "",
        ));
      } else if (child.tag === "load_step") {
        const component = attr(child, "component", "");
        setup.set(`load_step:${component}`, [
          attr(child, "from", ""),
          attr(child, "to", ""),
          attr(child, "at", "0s"),
          attr(child, "rise", "1us"),
        ].join(","));
      } else {
        setup.set(attr(child, "net", child.tag), attr(child, "value", ""));
      }
    }
    const assertions: Array<Record<string, string>> = [];
    for (const assertion of childElements(test)) {
      if (assertion.tag.startsWith("assert_")) {
        const a: Record<string, string> = {};
        for (const [k, v] of assertion.attrib) a[k] = v;
        a["op"] = assertion.tag;
        assertions.push(a);
      }
    }
    const run = find(test, "run");
    tests.set(testId, {
      id: testId,
      description: childText(test, "description"),
      setup,
      duration: run !== null ? attr(run, "duration", "") : "",
      assertions,
    });
  }

  const profiles = new Map<string, SimulationProfile>();
  for (const profile of findAllPath(root, "simulation_profiles", "profile")) {
    const profileId = attr(profile, "id", "");
    const props = new Map<string, string>();
    for (const p of findAll(profile, "property")) {
      const name = p.attrib.get("name");
      if (name) {
        props.set(name, attr(p, "value", ""));
      }
    }
    profiles.set(profileId, {
      id: profileId,
      default: attr(profile, "default", "false").toLowerCase() === "true",
      backends: findAll(profile, "backend").map((b) => attr(b, "type", "")),
      included_subsystems: findAll(profile, "include").map((i) =>
        attr(i, "subsystem", ""),
      ),
      tests: findAll(profile, "run").map((r) => attr(r, "test", "")),
      properties: props,
    });
  }

  const exports: ExportTarget[] = findAllPath(root, "exports", "export").map(
    (e) => ({
      target: attr(e, "target", ""),
      enabled: attr(e, "enabled", "false").toLowerCase() === "true",
    }),
  );

  return {
    name: attr(root, "name", ""),
    ir_version: attr(root, "ir_version", ""),
    metadata,
    requirements: [],
    nets,
    power_domains: powerDomains,
    components,
    interfaces,
    analog,
    firmware_projects: firmwareProjects,
    firmware_bindings: firmwareBindings,
    firmware_tasks: firmwareTasks,
    bridges,
    tests,
    simulation_profiles: profiles,
    exports,
  };
}

// --- helpers ---------------------------------------------------------------- #

function findAllPath(root: XmlElement, a: string, b: string): XmlElement[] {
  const out: XmlElement[] = [];
  for (const container of findAll(root, a)) {
    for (const el of findAll(container, b)) out.push(el);
  }
  return out;
}

/** Mirror of parser._text: stripped text of child `tag`, or "" (parent null). */
function childText(parent: XmlElement | null, tag: string): string {
  if (parent === null) return "";
  const child = find(parent, tag);
  if (child === null) return "";
  const text = elementText(child);
  return text ? text.trim() : "";
}

/** Mirror of parser._child_attr: attr `attrName` of child `tag`, or "". */
function childAttr(parent: XmlElement, tag: string, attrName: string): string {
  const child = find(parent, tag);
  return child !== null ? attr(child, attrName, "") : "";
}

/** dict(child.attrib): plain object of an element's attributes, ordered. */
function attribMap(el: XmlElement): Record<string, string> {
  const out: Record<string, string> = {};
  for (const [k, v] of el.attrib) out[k] = v;
  return out;
}

/**
 * Python `a or b or c` on attribute values: the first truthy (non-empty,
 * present) string wins; falls through to undefined so the caller can `?? ""`.
 * An empty-string attribute is falsy in Python, so it is skipped.
 */
function orChain(...values: Array<string | undefined>): string | undefined {
  for (const v of values) {
    if (v) return v;
  }
  return undefined;
}
